libraries/botbuilder/src/teamsActivityHelpers.ts

```diff
@@ -6,50 +6,59 @@
  */
 
-import { Activity, ChannelInfo, NotificationInfo, TeamInfo, TeamsChannelData } from 'botbuilder-core';
+import { Activity, TeamInfo, TeamsChannelData } from 'botbuilder-core';
 
-/**
- * Activity helper methods for Teams.
- */
-export function teamsGetChannelId(activity: Activity): string {
+function isTeamsChannelData(channelData: unknown): channelData is TeamsChannelData {
+    return typeof channelData === 'object';
+}
+
+function validateActivity(activity: Activity): void {
     if (!activity) {
         throw new Error('Missing activity parameter');
     }
-
-    const channelData: TeamsChannelData = activity.channelData as TeamsChannelData;
-    const channel: ChannelInfo = channelData ? channelData.channel : null;
-    return channel && channel.id ? channel.id : null;
 }
 
-export function teamsGetTeamId(activity: Activity): string {
-    if (!activity) {
-        throw new Error('Missing activity parameter');
+/**
+ * Activity helper methods for Teams.
+ */
+
+export function teamsGetTeamInfo(activity: Activity): TeamInfo | null {
+    validateActivity(activity);
+
+    const channelData = activity.channelData;
+    if (isTeamsChannelData(channelData)) {
+        const team = channelData.team;
+        return team || null;
     }
 
-    const channelData: TeamsChannelData = activity.channelData as TeamsChannelData;
-    const team: TeamInfo = channelData ? channelData.team : null;
+    return null;
+}
+
+export function teamsGetTeamId(activity: Activity): string | null {
+    const team = teamsGetTeamInfo(activity);
     return team && team.id ? team.id : null;
 }
 
-export function teamsNotifyUser(activity: Activity): void {
-    if (!activity) {
-        throw new Error('Missing activity parameter');
-    }
+export function teamsGetChannelId(activity: Activity): string | null {
+    validateActivity(activity);
 
-    if (!activity.channelData || typeof activity.channelData !== 'object') {
-        activity.channelData = {};
+    const channelData = activity.channelData;
+    if (isTeamsChannelData(channelData)) {
+        const channel = channelData.channel;
+        return channel && channel.id ? channel.id : null;
     }
 
-    const channelData: TeamsChannelData = activity.channelData as TeamsChannelData;
-    channelData.notification = { alert: true } as NotificationInfo;
+    return null;
 }
 
-export function teamsGetTeamInfo(activity: Activity): TeamInfo {
-    if (!activity) {
-        throw new Error('Missing activity parameter');
+export function teamsNotifyUser(activity: Activity, alertInMeeting?: boolean, externalResourceUrl?: string): void {
+    validateActivity(activity);
+
+    if (!isTeamsChannelData(activity.channelData)) {
+        activity.channelData = {};
     }
 
-    const channelData: TeamsChannelData = activity.channelData as TeamsChannelData;
-    const team: TeamInfo = channelData ? channelData.team : null;
-    return team;
+    if (isTeamsChannelData(activity.channelData)) {
```

Comment:
Does this check of isTeamsChannelData do anything?  line 57 looks like it will ensure there is a TeamsChannelData.

Reply:
It's a type guard, which means that inside the if block the Typescript compiler will understand that `activity.channelData` is, indeed, typed as TeamsChannlData. It's better practice to use type guards rather than casting.

Reply:
It is still confusing to me ... when will the 2nd one ever be false?

```js
    if (!isTeamsChannelData(activity.channelData)) {
        activity.channelData = {};
    }

    if (isTeamsChannelData(activity.channelData)) {
        activity.channelData.notification = { alert: true, alertInMeeting, externalResourceUrl };
    }
```

Not a big deal, I would just like to understand.

Reply:
It will never be false, but a side effect of the is check is that inside the block, where I'm accessing `activity.channelData`, the compiler (and thus intellisense) knows that the type of `activity.channelData` is `TeamsChannelData` rather than `any`. It's a convenient way to nudge the compiler that is more type safe than doing `(activity.channelData as TeamsChannelData)`. The handbook has some more to say about it [here](https://www.typescriptlang.org/docs/handbook/advanced-types.html).